velox/exec/RowContainer.cpp

```diff
@@ -437,6 +454,53 @@ void RowContainer::freeAggregates(folly::Range<char**> rows) {
   }
 }
 
+void RowContainer::freeNextRowVectors(folly::Range<char**> rows, bool clear) {
+  if (!nextOffset_ || !hasDuplicateRows_) {
+    return;
+  }
+
+  if (clear) {
+    for (auto row : rows) {
+      auto vector = getNextRowVector(row);
+      if (vector) {
+        // Clear all rows, we can clear the nextOffset_ slots and delete the
+        // next-row-vector.
+        for (auto& next : *vector) {
+          getNextRowVector(next) = nullptr;
+        }
+        auto allocator = vector->get_allocator().allocator();
```

Comment:
Any reason we cannot use `stringAllocator_` directly here?

Reply:
This is because currently the memory for the next row vectors may be not allocated from the RowContainer to which the row belongs. Please refer to the logic of 'parallelJoinBuild'. We need to allocate memory for the next row vectors in different threads from same RowContainers, and HashStringAllocator is not thread-safe. To address this issue, we allocate memory from the row container represented by the 'partition' parameter.
https://github.com/facebookincubator/velox/blob/main/velox/exec/HashTable.cpp#L946
https://github.com/facebookincubator/velox/blob/main/velox/exec/HashTable.cpp#L1034

I think there are two possible solutions to this issue: one is the current approach, and the other is the one I mentioned in the PR, which is to add a thread-safe StlAllocator for appendNextRow function. Please let me know your thoughts and any suggestions.

Reply:
The current approach is good, just add some comment to explain why we are using a different HSA.

Reply:
Thanks, added comments.